src/cmdstan/command_helper.hpp

```diff
@@ -0,0 +1,589 @@
+  std::vector<double> u_params_r;
+  std::vector<size_t> dims_u_params_r;
+  std::stringstream msg;
+
+  std::shared_ptr<stan::io::var_context> upars_context = get_var_context(fname);
+  u_params_r = upars_context->vals_r("params_r");
+  if (u_params_r.size() == 0) {
+    msg << "Unconstrained parameters file has no variable 'params_r' with "
+           "unconstrained parameter values!";
+    throw std::invalid_argument(msg.str());
+  }
+  // is input single vector of params or array of vectors?
+  dims_u_params_r = upars_context->dims_r("params_r");
+  u_params_rows = dims_u_params_r.size() == 2 ? dims_u_params_r[0] : 1;
+  u_params_cols
+      = dims_u_params_r.size() == 2 ? dims_u_params_r[1] : dims_u_params_r[0];
+  size_t num_upars = model.num_params_r();
+  if (u_params_cols != num_upars) {
+    msg << "Incorrect number of unconstrained parameters provided! "
+           "Model has "
+        << num_upars << " parameters but " << u_params_cols << " were found.";
+    throw std::invalid_argument(msg.str());
+  }
+  // brute force reshape
+  std::vector<std::vector<double>> params_r_ind(
+      u_params_rows, std::vector<double>(u_params_cols));
+  size_t idx = 0;
+  for (size_t i = 0; i < u_params_rows; ++i) {
+    for (size_t j = 0; j < u_params_cols; ++j) {
+      params_r_ind[i][j] = *(u_params_r.data() + idx);
+      ++idx;
+    }
+  }
+  return params_r_ind;
+}
+
+/**
+ * Extract and validate constrained parameter inputs.
+ * Input file contains definitions for all model parmeters.
+ *
+ * @param fname name of file which exists and has read perms.
+ * @param model Stan model
+ * @return Eigen vector of parameter estimates.
+ */
+std::vector<std::vector<double>> get_cparams_r(
+    const std::string &fname, const stan::model::model_base &model) {
+  std::vector<std::string> param_names;
+  std::vector<std::vector<size_t>> param_dimss;
+  stan::services::get_model_parameters(model, param_names, param_dimss);
+
+  size_t num_upars = model.num_params_r();
+  std::vector<double> params(num_upars);
+  std::vector<std::vector<double>> params_r_ind = {params};
+  std::vector<int> dummy_params_i;
+  std::shared_ptr<stan::io::var_context> cpars_context = get_var_context(fname);
+  std::vector<std::string> input_cpar_names;
+  cpars_context->names_r(input_cpar_names);
+  std::stringstream msg;
+  for (const std::string &m_param_name : param_names) {
+    if (!cpars_context->contains_r(m_param_name)) {
+      msg << "Constrained value(s) for parameter " << m_param_name
+          << " not found!";
+      throw std::invalid_argument(msg.str());
+    }
+  }
+  model.transform_inits((*cpars_context), dummy_params_i, params_r_ind[0],
+                        &msg);
+  return params_r_ind;
+}
+
+/**
+ * Read fitted params from sample in StanCSV format.
+ *
+ * @param fname name of file which exists and has read perms.
+ * @param model Stan model
+ * @return Eigen vector of parameter estimates.
+ */
+std::vector<std::vector<double>> get_cparams_r_csv(
```

Comment:
Since this is used in a call to log_prob_grad, the results here need to be put on to the unconstrained scale with `transform_inits` for them to be meaningful (c.f. the above comment and function)

Reply:
refactored